doc/src/manual/methods.md

```diff
@@ -515,6 +515,226 @@ julia> wait(schedule(t, 1))
+# For example, via primitive operations exposed by the processor.
++(a::Float, b::Float) = Core.add(a, b)
+```
+
+### Trait-based dispatch
+
+A natural extension to the iterated dispatch above is to add a layer to
+method selection that allows to dispatch on sets of types which are disjoint from
+the sets defined by the type hierarchy.
+We could construct such a set by writing out a `Union` of the types in question,
+but then this set would not be extensible.
+However, by expressing such a set as a "Holy-trait",
+it is possible to allow new type additions to the set.
+
+Such a trait is implemented by defining a generic function which
+computes a different singleton value for each trait-set to which the
+function arguments may belong to.  If this function is pure there is
+no impact on performance compared to normal dispatch.
+
+The example above glossed over the implementation details of [`map`](@ref) and [`promote`](@ref),
+since those both operate in terms of type traits.
+When iterating over a matrix, such as in the implementation of `map`,
+one important question is what order to use to traverse the data.
+When `AbstractArray` subtypes implement the `Base.IndexStyle` trait,
+other functions such as `map` can dispatch on this information to pick
+the best algorithm (see [Abstract Array Interface](@ref man-interface-array)).
+This means that each subtype does not need to implement a custom version of `map`,
+since the generic definitions + trait classes will enable the system to select the fastest version.
+
+```julia
+map(a::AbstractArray, b::AbstractArray) = map(Base.IndexStyle(a, b), a, b)
+map(::Base.IndexCartesian, a::AbstractArray, b::AbstractArray) = # generic implementation
+map(::Base.IndexLinear, a::AbstractArray, b::AbstractArray) = # linear-fast implementation
+```
+
+This trait-based approach is also present in the [`promote`](@ref) mechanism employed by `+`.
+It uses the types of arguments to deduce the optimal common type to
+compute the operation.
+This makes it possible to reduce the problem of implementing every function for every pair of possible type arguments,
+to the much smaller problem of implementing a conversion operation from each type to a common type,
+plus a table of preferred pair-wise promotion rules.
+
+
+### Output-type Computation
+
+The discussion of trait-based promotion provides a transition into our next design pattern:
+computing the output element type for a matrix operation.
+
+For implementing primitive operations, such as addition,
+we use the [`promote_type`](@ref) function to compute the desired output type.
+(As before, we saw this at work in the `promote` call in the call to `+`).
+
+For more complex functions on matrices, it may be necessary to compute the expected return
+type for a more complex sequence of operations.
+This is often performed by the following steps:
+
+1. Write a small function `op` that expresses the set of operations performed by the kernel of the algorithm.
```

Comment:
_Define a small helper function `op` that exposes the numerical operations performed by the algorithm_?